### ISSUE TITLE
Compiler panic in objrs package attribute macros class and root_class

### ISSUE THREAD

**Issue by ottobommer** [rust-lang/rust#82836]
### Code

git clone https://gitlab.com/ootoo1/objrs.git
cd objrs/frameworks/foundation/
cargo build

### Meta

`rustc --version --verbose`:
```
rustc 1.52.0-nightly (caca2121f 2021-03-05)
binary: rustc
commit-hash: caca2121ffe4cb47d8ea2d9469c493995f57e0b5
commit-date: 2021-03-05
host: x86_64-apple-darwin
release: 1.52.0-nightly
LLVM version: 12.0.0
```

### Error output

```
   Compiling objrs_frameworks_foundation v0.0.3-dev (/Users/ootoo/Projekte/objrs/frameworks/foundation)
error: any use of this value will cause an error
  --> frameworks/foundation/src/nsobject.rs:18:1
   |
18 | #[objrs(class, root_class)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ inbounds test failed: 0x0 is not a valid pointer
   |
   = note: `#[deny(const_err)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #71800 <https://github.com/rust-lang/rust/issues/71800>
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: any use of this value will cause an error
  --> frameworks/foundation/src/nsobject.rs:18:1
   |
18 | #[objrs(class, root_class)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ referenced constant has errors
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #71800 <https://github.com/rust-lang/rust/issues/71800>
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

thread 'rustc' panicked at 'called `Result::unwrap()` on an `Err` value: Unknown(T)', compiler/rustc_lint/src/types.rs:802:52
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.52.0-nightly (caca2121f 2021-03-05) running on x86_64-apple-darwin

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [lint_mod] linting module `nsarray`
#1 [analysis] running analysis passes on this crate
end of query stack
```

**Comment by ottobommer** [comment]
### Code
Minor changes to https://gitlab.com/ootoo1/objrs.git (added min_const_generics + allow(const_err)), the above errors are gone, but the compiler panic remains.

### Error output
```
warning: the feature `min_const_generics` has been stable since 1.51.0 and no longer requires an attribute to enable
 --> macros/src/lib.rs:6:73
  |
6 | #![feature(proc_macro_diagnostic, proc_macro_span, proc_macro_def_site, min_const_generics)]
  |                                                                         ^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(stable_features)]` on by default

warning: 1 warning emitted

   Compiling objrs_frameworks_foundation_macros v0.0.3-dev (/Users/ootoo/tmp/objrs/frameworks/foundation/macros)
warning: the feature `specialization` is incomplete and may not be safe to use and/or cause compiler crashes
  --> src/lib.rs:13:3
   |
13 |   specialization,
   |   ^^^^^^^^^^^^^^
   |
   = note: `#[warn(incomplete_features)]` on by default
   = note: see issue #31844 <https://github.com/rust-lang/rust/issues/31844> for more information
   = help: consider using `min_specialization` instead, which is more stable and complete

warning: 1 warning emitted

   Compiling objrs_frameworks_foundation v0.0.3-dev (/Users/ootoo/tmp/objrs/frameworks/foundation)
warning: field is never read: `isa`
  --> frameworks/foundation/src/nsobject.rs:22:3
   |
22 |   isa: *mut objrs::Class,
   |   ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

thread 'rustc' panicked at 'called `Result::unwrap()` on an `Err` value: Unknown(T)', compiler/rustc_lint/src/types.rs:802:52
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.52.0-nightly (caca2121f 2021-03-05) running on x86_64-apple-darwin

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [lint_mod] linting module `nsarray`
#1 [analysis] running analysis passes on this crate
end of query stack
warning: 1 warning emitted

error: could not compile `objrs_frameworks_foundation`

To learn more, run the command again with --verbose.
```

**Comment by hellow554** [comment]
I can't compile this code:

```
error: /tmp/tmp.XE6gfi1VBB/objrs/target/debug/deps/libobjrs_macros-214fa04ac9aaf3b1.so: undefined symbol: arc4random_buf
 --> src/__objrs.rs:8:1
  |
8 | pub extern crate objrs_macros;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

error: could not compile `objrs`
```

@ottobommer what is wrong here? 
I did exactly what you wrote (clone, cd, cargo build) with your exact compiler version.

**Comment by hellow554** [comment]
This is a somewhat minimal example, next step would be to get rid of all the other code (:

```
extern crate objrs;

use objrs::objrs;

#[objrs(class, root_class)]
pub struct NSArray<T: objrs::marker::Class + ?Sized>;

#[objrs(impl)]
#[link(name = "Foundation", kind = "framework")]
impl<T: objrs::marker::Class + ?Sized> NSArray<T> {
    #[objrs(selector = "firstObject")]
    pub fn first_object<'a>(&'a self) -> Option<&'a T> {}
}
```

**Comment by hellow554** [comment]
searched toolchains nightly-2020-06-01 through nightly-2021-02-04
Regression in nightly-2020-08-09

So the error is somewhere here: 09f4c9f5082f78b0adcee83d3ab4337e000cd28e...ceedf1d5febd65b012b8bcd513d70a0a6a091210

sadly there are no more artifacts available and I don't want to compile rust on my machine today.. ^^ So this must be it.

**Comment by ottobommer** [comment]
@apiraino OK, will try to make a smaller example. As it is not my code this may take some time.

**Comment by ottobommer** [comment]
I minimized the example somewhat, it can be pulled from the repository below, let me know if a further reduction is needed.
```
git clone https://gitlab.com/ootoo1/objrs-min.git
cd objrs-min/frameworks/foundation/
cargo build
    Updating crates.io index
  Downloaded syn v1.0.64
  Downloaded 1 crate (231.5 KB) in 0.48s
   Compiling proc-macro2 v1.0.24
   Compiling unicode-xid v0.2.1
   Compiling syn v1.0.64
   Compiling libc v0.2.88
   Compiling quote v1.0.9
   Compiling objrs_macros v0.0.3-dev (/Users/ootoo/tmp/objrs-min/macros)
   Compiling objrs v0.0.3-dev (/Users/ootoo/tmp/objrs-min)
   Compiling objrs_frameworks_foundation v0.0.3-dev (/Users/ootoo/tmp/objrs-min/frameworks/foundation)
thread 'rustc' panicked at 'called `Result::unwrap()` on an `Err` value: Unknown(T)', compiler/rustc_lint/src/types.rs:802:52
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.52.0-nightly (caca2121f 2021-03-05) running on x86_64-apple-darwin

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [lint_mod] linting top-level module
#1 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `objrs_frameworks_foundation`

To learn more, run the command again with --verbose.
```

**Comment by JohnTitor** [comment]
Assigning `P-medium` [as discussed as part of the Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2382836.20Compiler.20panic.20in.20objrs.20package.20attribute.20macros.20cla.E2.80.A6) and removing `I-prioritize`.

**Comment by hellow554** [comment]
@ottobommer an ideal testcase would be one file, but sometimes the proc macro is needed, thefore 1 + 1 for proc macro would work as well.

**Comment by ottobommer** [comment]
@hellow554 OK, I got it down to 3 files (proc macro lib.rs, lib.rs, test case panic.rs), however the total lines of code is still around 1000:
```
git clone https://gitlab.com/ootoo1/objrs-min.git
cd objrs-min/
cargo test
    Updating crates.io index
  Downloaded libc v0.2.89
  Downloaded 1 crate (516.0 KB) in 1.45s
   Compiling proc-macro2 v1.0.24
   Compiling unicode-xid v0.2.1
   Compiling syn v1.0.64
   Compiling libc v0.2.89
   Compiling quote v1.0.9
   Compiling objrs_macros v0.0.3-dev (/Users/ootoo/tmp/objrs-min/macros)
   Compiling objrs v0.0.3-dev (/Users/ootoo/tmp/objrs-min)
thread 'rustc' panicked at 'called `Result::unwrap()` on an `Err` value: Unknown(T)', compiler/rustc_lint/src/types.rs:802:52
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.52.0-nightly (caca2121f 2021-03-05) running on x86_64-apple-darwin

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [lint_mod] linting top-level module
#1 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `objrs`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by hellow554** [comment]
In the end, it's so simple...

```
extern "C" fn _panic<'a, T: ?Sized>() -> Option<&'a T> {
    panic!()
}

fn main() {}
```

It seems to be a problem with `?Sized` and `extern "C"`.

After all, my humble advice would be, stay away from that code. It is the most unrusty code I've seen in a long time. So many mut variables, returns, unnecessary nested types (Result<Result<...>>)... I mean it's impressive that it works, and if you have a usecase, go for it. Make the world a better place ;)

I'm bisecting it now.

@rustbot modify labels: -E-needs-mcve

**Comment by hellow554** [comment]
Regression in nightly-2020-07-31, somewhere between db0492ace429cfeb3567e2c04e300be7df9972ff...cfc572cae2d1fc381cce476b5c787fd7221af98c

**Comment by apiraino** [comment]
(not sure if it helps but) I'll add the relevant merged PRs
```
found 10 bors merge commits in the specified range
  commit[0] 2020-07-29UTC: Auto merge of #74733 - richkadel:llvm-coverage-map-gen-5, r=tmandry
  commit[1] 2020-07-29UTC: Auto merge of #73767 - P1n3appl3:rustdoc-formats, r=tmandry
  commit[2] 2020-07-30UTC: Auto merge of #74929 - Manishearth:rollup-z2vflrp, r=Manishearth
  commit[3] 2020-07-30UTC: Auto merge of #74923 - ehuss:update-cargo, r=Mark-Simulacrum
  commit[4] 2020-07-30UTC: Auto merge of #74876 - oli-obk:lumberjack_disable, r=RalfJung
  commit[5] 2020-07-30UTC: Auto merge of #74105 - npmccallum:naked, r=matthewjasper
  commit[6] 2020-07-30UTC: Auto merge of #74930 - ecstatic-morse:const-size-align-of-val, r=oli-obk
  commit[7] 2020-07-30UTC: Auto merge of #73990 - jumbatm:clashing-extern-decl, r=nagisa
  commit[8] 2020-07-30UTC: Auto merge of #74908 - RalfJung:miri, r=RalfJung
  commit[9] 2020-07-30UTC: Auto merge of #74957 - Manishearth:rollup-3wudwlg, r=Manishearth
```

**Comment by hellow554** [comment]
@jonas-schievink you tagged this as `A-Lint`, but I don't think it has anything to do with linting. May you be able to spot the right label for this?

**Comment by jonas-schievink** [comment]
I tagged it based on the panic path `compiler/rustc_lint/src/types.rs`.

The backtrace involves `rustc_lint::types::ImproperCTypesVisitor::check_type_for_ffi`, so this looks like a duplicate of https://github.com/rust-lang/rust/issues/80125

**Comment by jonas-schievink** [comment]
Ah, nevermind, that one looks related but different.

**Comment by hellow554** [comment]
https://github.com/rust-lang/rust/issues/52334 looks also very similar to this.

**Comment by Enselic** [comment]
Turns out it is not a duplicate, because this still reproduces on nightly. An easy way to reproduce is to copy-paste this to bash:

```sh
cat <<EOF | rustc +nightly --crate-type lib -
extern "C" fn _panic<'a, T: ?Sized>() -> Option<&'a T> {
    panic!()
}
EOF
```

**Comment by JohnTitor** [comment]
Triage: Fixed on the latest nightly, marking as `E-needs-test` but feel free to close if there's a regression test already.

**Comment by kadiwa4** [comment]
Fixed in 1.74.0 by #115631, which already added a test to `tests/ui/lint/lint-ctypes-option-nonnull-unsized.rs`.
